### PR TITLE
add `ENABLED` and `\cdot` front-end

### PR DESCRIPTION
**Pull request by johnyf**
These changes:

- add the following proof directives to the module `library/TLAPS.tla`:
  - `ENABLEDaxioms`: apply proof rules about the operator `ENABLED` (implemented in the module `src/expr/e_action.ml`), and ensure for soundness that assumptions are of appropriate expression level (implemented in the module `src/module/m_elab.ml`)
  - `ExpandENABLED`: replace the operator `ENABLED` with rigid quantification of primed occurrences of variables (implemented in the module `src/expr/e_action.ml`)
  - `ExpandCdot`: replace the operator `\cdot` with rigid quantification over primed and unprimed occurrences of variables (implemented in the module `src/expr/e_action.ml`)
  - `AutoUSE`: expand defined operators where needed to soundly replace the operators `ENABLED` and `\cdot` with rigid quantification, raise errors for declared operators and parameters of operators (implemented in the module `src/expr/e_action.ml`)
  - `Lambdify`: replace the operator `ENABLED` and `\cdot` with constant-level expressions that include `LAMBDA` expressions with normalized parameter names (implemented in the module `src/expr/e_action.ml`)
  - `LevelComparison`: prove sequents with modified levels of declared operators, for example replace an `ACTION` declaration with a `STATE` declaration (implemented in the module `src/expr/e_level_comparison.ml`)
- computation of TLA+ expression levels (implemented in the module `src/expr/e_levels.ml`)
- add version 13 of the fingerprints file format, which includes the newly implemented proof directives (implemented in the module `src/backend/fpfile.ml`)
- add command-line option `--line <int>` for line to prove (abbreviates `--toolbox n n`) (implemented in the module `src/tlapm_args.ml`)
- add command-line option `--smt-logic <logic>` for setting the SMT solver logic, with default the logic UFNIA (implemented in the modules `src/backend/smt.ml`, `src/params.ml`, `src/tlapm_args.ml`)
- save fingerprints and auxiliary files that correspond to the file `filename.tla` in the directory `__tlacache__/filename.tlaps` (implemented in the module `src/tlapm.ml`, updated test file `test/fast/fingerprint/load_v8old_test.tla`)
- save kind of declaration (`CONSTANT`, `STATE`, `ACTION`, `TEMPORAL`, unknown) in fingerprints (implemented in the module `src/backend/fingerprints.ml`)
- save proof directives that appear in `BY` statements in the fingerprints, to ensure that proof obligations with the same assertions and different `BY` statements result in different fingerprints (implemented in the module `src/backend/fingerprints.ml`)
- add tests for the newly implemented proof directives to the directory `test/fast/enabled_cdot/`:
  - `Cdot_test.tla`
  - `ENABLEDaxioms_test.tla`
  - `Enabled_test.tla`
  - `ExpandENABLED_LET_test.tla`
  - `ExpandOnlyCdot_test.tla`
  - `ExpandOnlyENABLED_test.tla`
  - `Level_test.tla`
  - `NestedCdot_test.tla`
  - `NestedENABLED_from_AutoUSE_test.tla`
  - `NestedENABLED_test.tla`
- bug fixes to coalescing (in the module `src/frontend/coalesce.ml`)
- prepend `CONSTANT_`, `STATE_`, `ACTION_`, `TEMPORAL_`, `Unknown_`, `VARIABLE_` to coalesced identifier, to represent level information about declared and defined operators and avoid related unsoundness bugs (implemented in `src/frontend/coalesce.ml`)
- reimplement substitutivity computation (in the module `src/expr/e_substitutive.ml`)
- add class `map` to module `src/expr/e_subst.ml` (to enable subclassing)
- add class `map_visible_hyp` to module `src/expr/e_subst.ml` (substitution that visits only visible hypotheses in a sequent's context)
- add classes `map_visible_hyp`, `iter_visible_hyp`, `map_rename` to the module `src/expr/e_visit.ml`
- add functions `set_to_list`, `collect_unprimed_vars`, `collect_primed_vars` to the module `src/expr/e_visit.ml`
- call function `map_visible_hyp` from module `src/frontend/action.ml`
- call level computation in `src/frontend/action.ml`, instead of `Expr.Constness.is_const`
- bug fix in `src/frontend/action.ml` to add `ENABLED` in pattern case
- bug fix in `src/frontend/action.ml` to assert false for the operators `UNCHANGED`, `\cdot`, `~>`, `-+->`, `[]`, `<>`, `\AA`, `\EE`, subscript, temporal subscript, `WF_`, `SF_` in pattern case
- add comments to the types in the module `src/expr/e_t.ml`
- add functions `visibility_to_string`, `print_cx`, `cx_front`, `scx_front`, `find_hyp_named`, `expr_locus`, `format_locus`, `shape_to_arity`, `number_of_hyp`, `sequent_stats` to the module `src/expr/e_t.ml`
- add the example of incrementing an integer-valued variable `examples-draft/SimpleExampleWF.tla`
- activate proving for submodules (implemented in the module `src/module/m_parser.ml`)
- modules from `INSTANCE` and `EXTENDS` statements could be submodules of a module that the module extends, and is itself another file (implemented in the module `src/module/m_dep.ml`, `src/module/m_save.ml`)
- refactor module `src/module/m_elab.ml`, part moved to the module `src/module/m_subst.ml` (described below)
- add visitor of TLA+ module syntax graphs (implemented in the module `src/module/m_visit.ml`)
- add utilities for performing substitutions in TLA+ module syntax graphs (implemented in the module `src/module/m_subst.ml`)
- refactor `src/tlapm.ml`, change error messages, printing of count of failed obligations
- `failwith_unsupp` for the operators `~>`, `ENABLED`, `\cdot`, `-+->`, `[]`, `<>` in the Isabelle and Zenon backends (implemented in the modules `src/backend/isabelle.ml`, src/backend/zenon.ml)
- coalesce defined operator occurrences, operator application, and first-order quantification in `src/backend/ls4.ml`
- update the modules `src/backend/prep.ml`, `src/method.ml`, `src/method_prs.ml` with the newly implemented proof directives
- add `~>` to temporal operators in pattern case in the module `src/expr/e_elab.ml`
- add function `first_n` to module `src/util/deque.ml`
- augment error message in the module `src/expr/e_deref.ml`
- `BSeq` was a binary operator (implemented in the module `src/module/m_standard.ml`)
- update the documentation file `todo.txt`
- update the file `src/Makefile`
- update the generated file `src/.depend`

### CODE REVIEW

**Comment by kape1395** [comment]
It fails to compile for me with error:
```
File "e_visit.ml", line 470, characters 14-33:
Error: Unbound value Hashtbl.to_seq_keys
```

**Comment by kape1395** [comment]
It works with ocaml 4.09.1 and fixes #8, at least for my case.

**Review by damiendoligez — COMMENTED** [review]
Looks good in general. I have a commented a few points of detail.

**Comment by johnyf** [comment]
Thank you for reviewing, I will address the comments.

File: src/expr/e_levels.ml [hunk 1]
```diff
@@ -0,0 +1,1073 @@
...
+            let level = max e_level doms_level in
+            let level_args = StringSet.union
+                e_level_args doms_level_args in
+            let level_info = LevelInfo (level, [], level_args) in
+            let new_fcn = Fcn (bs_, e_) in
+            let new_e = new_fcn @@ e in
+            assign new_e exprlevel level_info
+        | FcnApp (f, es) ->  (* f[x] *)
+                (* function `f` *)
+            let fes = List.append [f] es in
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/expr/e_levels.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_levels.ml [hunk 2]
```diff
@@ -0,0 +1,1073 @@
...
+            let (level_info, fes_) = level_info_from_args scx fes in
+            let f_ = List.hd fes_ in
+            let es_ = List.tl fes_ in
+            let new_fcnapp = FcnApp (f_, es_) in
+            let new_e = new_fcnapp @@ e in
+            assign new_e exprlevel level_info
+        | Arrow (a, b) ->
+            let args = [a; b] in
+            let (level_info, args_) = level_info_from_args scx args in
+            let a_ = List.hd args_ in
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/expr/e_levels.ml:747]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml [hunk 3]
```diff
@@ -400,3 +412,284 @@ class virtual ['s] iter = object (self : 'self)
...
+          let h = Flex s @@ h in
+          (adj scx h, h)
+      | Defn (_, _, Hidden, _)
+      | Fact (_, Hidden, _) ->
+        (* TODO: what about mutable properties of `h` ? *)
+        (adj scx h, h)
+      | Defn (df, wd, Visible, ex) ->
+          let h = Defn (self#defn scx df, wd, Visible, ex) @@ h in
+          (adj scx h, h)
+      | Fact (e, vis, tm) ->
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml [hunk 4]
```diff
@@ -400,3 +412,284 @@ class virtual ['s] iter = object (self : 'self)
...
+        method expr (((prime_scope: bool), cx) as scx) e =
+            match e.core with
+            | Apply ({core=Internal Prime}, _) ->
+                assert (not prime_scope)
+            | Ix n ->
+                begin
+                assert (n >= 1);
+                assert (not prime_scope);
+                let hyp = E_t.get_val_from_id cx n in
+                match hyp.core with
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml [hunk 5]
```diff
@@ -400,3 +412,284 @@ class virtual ['s] iter = object (self : 'self)
...
+                    | Flex name ->
+                        let var_name = name.core in
+                        if (not (Hashtbl.mem
+                                unprimed_vars var_name)) then
+                            Hashtbl.add unprimed_vars var_name ()
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/frontend/coalesce.ml [hunk 6]
```diff
@@ -14,110 +17,364 @@ open Expr.Subst
...
-  in
-  "h" ^ String.sub (Digest.to_hex (Digest.string s)) 0 30
+    let s =
+        let b = Buffer.create 10 in
+        let fmt = Format.formatter_of_buffer b in
+        Expr.Fmt.pp_print_expr (scx, cx) fmt e ;
+        Format.pp_print_flush fmt () ;
+        Buffer.contents b
+    in
+    "h" ^ String.sub (string_to_hex s) 0 5
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/frontend/coalesce.ml]: Thank you for the suggestion, I added it to the changes.

File: src/frontend/coalesce.ml [hunk 7]
```diff
@@ -14,110 +17,364 @@ open Expr.Subst
...
-  | m -> m :: (int_compr ls (m-1))
+  | m -> m :: (int_compr ls (m - 1))
+
+
+let indexed_args args =
+    (* Combine each item in `args` with an index to form the list
+    `[(n, args[0]); (n - 1, args[1]); ..; (1, args[n - 1])]`
+    *)
+    let n = List.length args in
+    List.combine (int_compr [] n) args
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/frontend/coalesce.ml]: Thank you for the suggestion, I added it to the changes. (It appears that `args` as the second argument to the function `List.mapi` is needed.)

File: src/system.ml [hunk 8]
```diff
@@ -109,13 +109,13 @@ let make_line_buffer fd = {desc = fd; buf = Buffer.create 100};;
 let rawbuf = Bytes.create 1000;;
 
 let read_lines buf =
-  let n = Unix.read buf.desc rawbuf 0 (String.length rawbuf) in
+  let n = Unix.read buf.desc rawbuf 0 (String.length (Bytes.to_string rawbuf)) in
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/system.ml]: Thank you for the suggestion, I added it to the changes.

File: src/system.ml [hunk 9]
```diff
@@ -109,13 +109,13 @@ let make_line_buffer fd = {desc = fd; buf = Buffer.create 100};;
...
   if n = 0 then begin
     let last = Buffer.contents buf.buf in
     Buffer.reset buf.buf;
     if last = "" then [Leof] else [Line last; Leof]
   end else begin
-    Buffer.add_substring buf.buf rawbuf 0 n;
+    Buffer.add_substring buf.buf (Bytes.to_string rawbuf) 0 n;
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/system.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/deque.ml [hunk 10]
```diff
@@ -105,10 +105,37 @@ let rec nth ?(backwards=false) q n =
...
+            failwith "Deque.nth: internal error"
     in
     git q n
 
+let rec first_n q n =
+    (* Return the first `n` elements of the queue `q`. *)
+    if (n < 0) then
+        failwith ("Deque.first_n:  n = " ^
+            (string_of_int n)
+            ^ " < 0");
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/util/deque.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/deque.ml [hunk 11]
```diff
@@ -105,10 +105,37 @@ let rec nth ?(backwards=false) q n =
...
+    if (n > (size q)) then
+        failwith ("Deque.first_n:  n = " ^
+            (string_of_int n) ^ " > size q = " ^
+            (string_of_int (size q)) );
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/util/deque.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/ext.ml [hunk 12]
```diff
@@ -134,12 +134,11 @@ module Std = struct
   let input_all : ?bsize:int -> in_channel -> string =
     fun ?(bsize = 19) ch ->
       let buf = Buffer.create bsize in
-(*      let str = String.create 64 in *)
       let str = Bytes.create 64 in
       let rec loop () =
         let hmany = Pervasives.input ch str 0 64 in
         if hmany > 0 then begin
-          Buffer.add_substring buf str 0 hmany ;
+          Buffer.add_substring buf (Bytes.to_string str) 0 hmany ;
```
[the remark this answers is not in the document]
Comment (johnyf) [reply src/util/ext.ml]: Thank you for the suggestion, I added it to the changes.

**Review by johnyf — COMMENTED** [review]
I have added a few more commits to address the comments.

**Review by damiendoligez — APPROVED** [review]
This all looks good now. Thanks for the changes and explanations.

**Comment by damiendoligez** [comment]
This is good to merge now. @johnyf I'll let you push the merge button after making any last-minute changes, if you have any.

**Comment by johnyf** [comment]
Thank you for approving these changes, I merged them.

**Pull request merged**